### ISSUE TITLE
Support command line parameters to add an additional layer of configurability in the documentation/code generator.

### ISSUE THREAD

**Issue by lquerel** [open-telemetry/weaver#193]
In #192, Liudmila expresses the need to make the generation of incubating assets optional while simultaneously reducing the level of redundant definitions in the templates. This issue is a proposal to solve those problems.

Let’s start by introducing two new arguments to the weaver registry generate command: `--param` and `--params`.

To specify one or several template parameters directly from the command line:
`weaver registry generate --param incubating=true --param excluded="[ios, jvm]" <target> <output-dir>`

To specify a YAML file containing a set of parameters:
`weaver registry generate --params <yaml-params-file> <target> <output-dir>`

The structure of this YAML file is as follows:
```yaml
params:
  incubating: true
  excluded: [ios, aspnetcore, signalr, android, dotnet, jvm]
```

> Note: the value of a param can be any valid YAML value.

In the `weaver.yaml`, we will also introduce a new optional section named `params`. The parameters provided via the command line will override the ones defined in the `weaver.yaml` params section. This way, we can provide default values for params in the `weaver.yaml` file.

A new global variable will be introduced in the evaluation context of the templates. We currently have two global variables, `ctx` and `template`. We will introduce a new one named `params`, which will expose the parameters defined in `weaver.yaml` and potentially overridden via the command line.

These parameters will be directly available in the templates, and conditional logic can be used to influence the generation of code or documentation.

```
...
{% if params.incubating %}
... generate incubating code ...
{% endif %}
...
```

The JQ expression has also access to the parameters via the $var notation. In the following example, the `$excluded` JQ global variable is defined in the params. 

```yaml
templates:
  - pattern: <glob-pattern>
    filter: >
      if $incubating then
        .groups
          | map(select(.type == "attribute_group"))
          | map(select(.id | startswith("registry")))
          | map({ id: .id, group_id: .id | split(".") | .[1], attributes: .attributes })
          | group_by(.group_id)
          | map({ id: .[0].group_id, attributes: [.[].attributes[]] | sort_by(.id), output: "_incubating/attributes/", stable_package_name: "opentelemetry.semconv.attributes" })
          | map(select(.id as $id | any($excluded[]; . == $id) | not))
          | map(select(.attributes | length > 0))
      else
        empty
      end
     application_mode: single | each
```

In the context of #192 and with this improvement in place, it will be easy to create a single target `python` to generate either stable artifacts or both `stable` and `incubating` artifacts based on the parameters provided in the command line.

**Comment by lmolkova** [comment]
Thank you for the detailed proposal!

It'd be great to make these params accessible in the ctx (in JQ and in templates). Then it's not really necessary to add `if` in the `pattern` section.

Examples:
1. I want to configure a list of excluded namespaces (e.g. `dotnet` and `jvm` are not relevant on python). I do something like `weaver registry generate --param excluded="dotnet, jvm"`. Then I can either build a jq filter or have some logic in the template to exclude such namespaces.
2. I want to generate stable and incubating artifacts. I can run weaver twice, but with different param values `weaver registry generate --param semconv_filter="stable"` and ``weaver registry generate --param semconv_filter="any"` then again, I can build a JQ filter or some tempate logic.

This would result in a very flexible solution. It's also not a problem to run weaver multiple times - the semconv generation is usually a script involving multiple steps anyway.

**Comment by lmolkova** [comment]
Note: this is an existing feature of build-tools 

e.g. Java passes a bunch of additional context properties:

```kotlin
        "-Dexcluded_namespaces=\"ios aspnetcore signalr\"",
        "-Dfilter=${filter}",
        "-DclassPrefix=${classPrefix}",
        "-Dpkg=$packageNameArg",
        "-DstablePkg=$stablePackageNameArg"))
```
https://github.com/open-telemetry/semantic-conventions-java/blob/756a10d3c3efcf7d87b2ac246e22d64cc394ddf6/build.gradle.kts#L104

It allows to keep all the codegen config in the script and avoid hardcoding things in the templates/jq filters or weaver.yaml

**Comment by lquerel** [comment]
@lmolkova Yes, the parameters will be visible from the Jinja templates and from the JQ expressions.

The `if` expression is optional, but when used, it's a bit more performant than a full evaluation of the entire template. More importantly, this approach is:
1. Easier to interpret because it's not buried within the templates themselves. It's directly visible in the config file.
2. A simple solution to avoid the problem of empty template evaluation if the logic inside the template ends up with an empty result.

Nothing prevents running Weaver multiple times with different targets and parameters. However, keep in mind that defining a single target and using parameters to configure exactly what you need will be faster, as all the tuples {template, JQ result} are evaluated in parallel. 

In your specific context, I recommend defining a single target, let's say `python`, and using parameters to determine whether to generate the incubating assets or not.

**Comment by lquerel** [comment]
The `-Dparam_name=param_value` will also be supported in addition to `--param param_name=param_value`

**Comment by jsuereth** [comment]
So - I think the updated proposal meets @lmolkova's requirements.  My only complaint right now is this:

```
templates:
  - pattern: <glob-pattern>
     filter: <jq-expression>
     if: <jinja-expression>
     application_mode: single | each
```

There are *three* different languages here, two of which aren't quite "common as sql".  I'm not saying these are bad choices, but I think debugability of the `if` jinja expression will be hard without some care/thought into output (e.g. can't just throw down a `debug(ctx)` to see what's happening in the template).

I'd like to prioritize:

- Leveraging common playgrounds or debugging tools where possible.
  - Easily setting up a debug environment (e.g. `registry generate` for JSON -> jq playground).
  - Leveraging common conventions (e.g. can copy-paste most jq or jinja snippets from examples).
- Iteratively improving "DRY" (don't repeat yourself) in weaver config.
- Flexibility

I'm with liudmila, I think `if` could wait for a bit before we commit, but params in JQ and then the rendered templates would be great - we should commit to that in the next release.

TL;DR; Love the proposal - think we just need the params in templates + jq expressions, no if statement.

**Comment by lquerel** [comment]
@lmolkova @jsuereth Based on your feedback, I removed the `if` statements in the `templates` section. It's indeed redundant with the JQ expression where we can easily specify an if/else statement based on the parameters. I have updated the `weaver.yaml` example to integrate this `if` directly into the JQ expression.

**Comment by lmolkova** [comment]
thank you @lquerel !
